### PR TITLE
kernel: renamespace and automatically document arch API

### PR DESCRIPTION
**Pull request by tyrantlizard77**
We would like to have automatically generated documentation for the interfaces between the core kernel and architecture implementations. This set of APIs is currently namespaced as z_arch_*. Promote it to a new namespace arch_*, and enable automatic documentation generation.

Some ARC-specific APIs have been appropriately renamed.

Some additional defines enabled to documentation generation to expose more APIs to the doc scanner.

### CODE REVIEW

**Review by dbkinder — APPROVED** [review]
doc changes LGTM.

**Review by dcpleung — APPROVED** [review]

**Review by wentongwu — APPROVED** [review]

**Review by aescolar — COMMENTED** [review]
About the posix arch/boards changes no complaint per se.
But I repeat my usual remark: This is an interface between Zephyr modules, and we should, in general, try to minimize the amount of changes we do to it (to avoid unnecessary trouble for off-tree versions, as well as extra work for other maintainers of in tree components)

**Comment by andrewboie** [comment]
> But I repeat my usual remark: This is an interface between Zephyr modules, and we should, in general, try to minimize the amount of changes we do to it (to avoid unnecessary trouble for off-tree versions, as well as extra work for other maintainers of in tree components)

@aescolar What is it that you want here?? These APIs are being promoted to non-private. This will make supporting out-of-tree code *easier*

I'm getting the impression that to you, all APIs are non-private and that simply isn't the case in Zephyr. Anything prefixed with z_/Z_ can change at *any* time, full stop.

**Comment by aescolar** [comment]
Hi @andrewboie, 

> What is it that you want here??

I did not request any change. And I am not blocking the PR.

What I wanted is to reiterate what I stated in the past: That internal APIs in between Zephyr subsystems, even that they are not considered public and should not follow a deprecation policy, should still be changed with care.
Even though me repeating this remark was triggered by this PR, it is not about the PR content itself but about the general principle. To make others aware that "_Anything prefixed with z/Z can change at any time, full stop_" is not a good enough policy, if it leads to an interface between submodules being changed all too often.

I do acknowledge though, that it may be a bit confusing the fact that I stated this general remark as a comment to this PR. Nevertheless I hope you agree that other members of the project should have some way of presenting their general concerns to you or others.
And as it is very rare for you and me to be present in the same project meetings, it seemed appropriate to communicate with you here in that form.

> These APIs are being promoted to non-private. This will make supporting out-of-tree code easier

Thank you for that. I had never asked for that much, but I certainly welcome it.

**Comment by nashif** [comment]
@andrewboie is arch_ being promoted to a public API here? Or does it remain internal with a different namespace?

**Review by nashif — APPROVED** [review]

**Comment by aescolar** [comment]
> we should maybe promote this and similar APIs to a managed, documented and internal contract and treat it differently from z_/Z_ APIs, something this PR is doing already.

Agreed

File: include/arch/arm/cortex_m/mpu/arm_core_mpu.h
Comment (ioannisg) [inline include/arch/arm/cortex_m/mpu/arm_core_mpu.h]: Could you see if these can be cleaned up? At some point I might have thought that this api would be cross-ARCH but it does not seem to be the case...

**Review by ioannisg — APPROVED** [review]
ARM changes look good; I've left a comment about cleaning up, possibly, some declarations.

**Comment by ioannisg** [comment]
@nashif I believe the arch_ should signify a non-private api...

**Comment by ioannisg** [comment]
Could we get some ARC guys to take a look? @vonhust FYI :)

**Comment by vonhust** [comment]
@andrewboie If I recall correctly,  the API changed from _arch to z_arch, and this time to arch_xxx.
So are changes in this PR the final version?  I agreed that the API should be improved continually for better maintainence, bettter abstract and easier understanding,  but we should try not to bring much confusion, especially for application/driver developers.

**Review by vonhust — APPROVED** [review]
LGTM if CI build is OK

**Review by ruuddw — APPROVED** [review]

File: include/arch/arm/cortex_m/mpu/arm_core_mpu.h
Comment (ioannisg) [reply include/arch/arm/cortex_m/mpu/arm_core_mpu.h]: I think the z_arm_ already exists, and is under arch/arm/core/cortex-m

Comment (ioannisg) [reply include/arch/arm/cortex_m/mpu/arm_core_mpu.h]: I think the z_arm_ already exists, and is under arch/arm/core/cortex-m

Comment (ioannisg) [reply include/arch/arm/cortex_m/mpu/arm_core_mpu.h]: I think the z_arm_ already exists, and is under arch/arm/core/cortex-m

Comment (nashif) [reply include/arch/arm/cortex_m/mpu/arm_core_mpu.h]: broken record?

Comment (ioannisg) [reply include/arch/arm/cortex_m/mpu/arm_core_mpu.h]: > broken record?

Something with my browser/GH account I guess. Sorry for the noise

**Comment by andrewboie** [comment]
> @andrewboie If I recall correctly, the API changed from _arch to z_arch

This happened quite some time ago, when *all* private APIs were renamed (_ to z_).

> and this time to arch_xxx. So are changes in this PR the final version?

These APIs are now treated as public, subject to Zephyr's deprecation policy, so yes.
That's what this patch does: it makes these APIs public. They are now treated as contracts.

**Comment by ioannisg** [comment]
Looks like a rebase is needed, but after that it could be merged.

**Pull request merged**
